### PR TITLE
Add Nice option to service

### PR DESCRIPTION
**Pull request by jorpilo**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x]  Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x]  Have you added an explanation of what your changes do and why you'd like us to include them?
- [x]  Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x]  Have you successfully run brew style with your changes locally?
- [x]  Have you successfully run brew typecheck with your changes locally?
- [x]  Have you successfully run brew tests with your changes locall

-----

Added service.nice to allow to define default nice levels in plist and systemd units.

### CODE REVIEW

File: Library/Homebrew/service.rb [hunk 1]
```diff
@@ -355,6 +355,20 @@ def command
       @run&.map(&:to_s)
     end
 
+    sig { params(value: T.nilable(Integer)).returns(T.nilable(Integer)) }
```
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:358]: Why/when would you want to pass in a `value: nil`?
```suggestion
    sig { params(value: Integer).returns(T.nilable(Integer)) }
```
may make more sense if not

File: docs/Formula-Cookbook.md [hunk 2]
```diff
@@ -891,7 +891,7 @@ There are two ways to add `launchd` plists and `systemd` services to a formula,
 This table lists the options you can set within a `service` block. Only the `run` field is required which indicates what to run.
 
 | method                  | default      | macOS | Linux | description |
-| ----------------------- | ------------ | :---: | :---: | ----------- |
+|-------------------------| ------------ | :---: |:-----:| ----------- |
```
Comment (MikeMcQuaid) [inline docs/Formula-Cookbook.md:894]:
```suggestion
| ----------------------- | ------------ | :---: | :---: | ----------- |
```

just leave this as-is, thanks

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the PR. Looking good so far!

File: Library/Homebrew/service.rb (hunk 1 above)
Comment (jorpilo) [reply Library/Homebrew/service.rb:358]: Correct, we can remove nil but I was following the format of the rest of parameters like restart_delay which is also defined nilable
https://github.com/Homebrew/brew/blob/0d70ca7ab0d9a669e85e50bcfedca7cc9c7135c3/Library/Homebrew/service.rb#L211

Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:358]: CC @SMillerDev for thoughts here, I think it'd probably make sense to make none of these nilable.

Comment (reitermarkus) [reply Library/Homebrew/service.rb:358]: Most of these methods simply return the current value when `nil` (i.e. no argument) is passed. We don't seem to be calling any of these without an argument though.

**Comment by MikeMcQuaid** [comment]
Thanks @jorpilo! Run `brew typecheck` locally and fix the error: https://github.com/Homebrew/brew/actions/runs/4379299101/jobs/7665097153#step:7:12

Essentially it's saying the `Integer` check here is no longer needed. If you're willing: would be good to clean up the rest of the file in this way, too, after this PR is merged!

File: Library/Homebrew/service.rb [hunk 3]
```diff
@@ -355,6 +355,16 @@ def command
       @run&.map(&:to_s)
     end
 
+    sig { params(value: Integer).returns(T.nilable(Integer)) }
+    def nice(value)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/service.rb:359]:
```suggestion
    def nice(value = nil)
```

Sorry, this bit is still needed and then then `when nil` below will also be needed. 

We want users to be able to call `nice()` or `nice(5)` but not `nice("5")` or `nice(nil`).

Does that make sense?

Sorry for all the back and forth!

File: Library/Homebrew/service.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:358]: @dduugg @reitermarkus Also would welcome your thoughts on how, in a future PR, we could refactor this type handling to lean more on Sorbet (bearing in mind that e.g. formulae will almost never be evaluated at runtime to use Sorbet and are currently never type checked by Sorbet)

File: Library/Homebrew/service.rb [hunk 4]
```diff
@@ -391,6 +408,7 @@ def to_plist
       base[:LegacyTimers] = @macos_legacy_timers if @macos_legacy_timers == true
       base[:TimeOut] = @restart_delay if @restart_delay.present?
       base[:ProcessType] = @process_type.to_s.capitalize if @process_type.present?
+      base[:Nice] = @nice if @nice.present?
```
Comment (dduugg) [inline Library/Homebrew/service.rb:411]:
```suggestion
      base[:Nice] = @nice if @nice
```
`present?` is generally unnecessary redirection, unless whitespace is concerned. (It's either `nil` or `Integer` if we get here, correct?)

File: Library/Homebrew/service.rb [hunk 5]
```diff
@@ -459,6 +477,7 @@ def to_systemd_unit
 
       options << "Restart=always" if @keep_alive.present? && @keep_alive[:always].present?
       options << "RestartSec=#{restart_delay}" if @restart_delay.present?
+      options << "Nice=#{@nice}" if @nice.present?
```
Comment (dduugg) [inline Library/Homebrew/service.rb:480]:
```suggestion
      options << "Nice=#{@nice}" if @nice
```

File: Library/Homebrew/service.rb (hunk 1 above)
Comment (dduugg) [reply Library/Homebrew/service.rb:358]: I don't think i have any especially creative suggestions. You can try to bump these files to `typed: strict` to catch more edge cases. I believe we do runtime typechecking in tests, so covering every `service` DSL in tests could do the trick. A rubocop that flags unacceptable`nice` values could help too (it's tricky if you're allowing vars to be passed to `nice`, but it could enforce that `nice` is followed by a literal `Integer` in the required range).

File: Library/Homebrew/service.rb [hunk 6]
```diff
@@ -355,6 +355,23 @@ def command
       @run&.map(&:to_s)
     end
 
+    sig { params(value: T.nilable(Integer)).returns(T.nilable(Integer)) }
+    def nice(value = nil)
+      case T.unsafe(value)
+      when nil
+        @nice
+      when Integer
+        if (T.must(value) < -20) || (T.must(value) > 19)
```
Comment (dduugg) [inline Library/Homebrew/service.rb:364]:
```suggestion
        unless (-20..19).cover?(nice)
```
This is concise, and also lets you avoid the `T.must`s. (You could even extract the magic numbers into a const, e.g. `NICE_RANGE = -20..19`)

File: Library/Homebrew/service.rb (hunk 3 above)
Comment (reitermarkus) [reply Library/Homebrew/service.rb]: When would be want to call `nice` without arguments though?

Comment (reitermarkus) [reply Library/Homebrew/service.rb]: I don't really see a reason why any of these methods would need to behave like a getter.

File: Library/Homebrew/service.rb (hunk 6 above)
Comment (carlocab) [inline Library/Homebrew/service.rb:364]: Do you need `sudo` for negative `nice`?

**Comment by jorpilo** [comment]
Also sorry for the late reply. I had all my comments as "pending" and I didn't realised I do need to accept the comments for them to be visible....

File: Library/Homebrew/service.rb (hunk 1 above)
Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:360]: Thanks @dduugg!

File: Library/Homebrew/service.rb [hunk 7]
```diff
@@ -459,6 +472,7 @@ def to_systemd_unit
 
       options << "Restart=always" if @keep_alive.present? && @keep_alive[:always].present?
       options << "RestartSec=#{restart_delay}" if @restart_delay.present?
```
[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/service.rb:481]: @SMillerDev can confirm

File: Library/Homebrew/service.rb (hunk 1 above)
Comment (SMillerDev) [reply Library/Homebrew/service.rb:360]: > CC @SMillerDev for thoughts here, I think it'd probably make sense to make none of these nilable.

Not a clue, pretty sure I just copied this from livecheck

File: Library/Homebrew/service.rb (hunk 6 above)
Comment (carlocab) [reply Library/Homebrew/service.rb]: I guess a warning when you call `brew services` without `sudo` suffices here. But CC @SMillerDev and @MikeMcQuaid for thoughts here.

Comment (SMillerDev) [reply Library/Homebrew/service.rb]: Since this is not something that should happen I think throwing an error is the best course of action.

Comment (carlocab) [reply Library/Homebrew/service.rb]: It seems reasonable to truncate the `nice` level if you don't want to run `brew services` with `sudo` (which is what will happen automatically), though, so throwing an error seems a bit much here.

Comment (SMillerDev) [reply Library/Homebrew/service.rb]: I believe `requires_root` still allows you to run as a user, it'll just warn against doing so. So if we require that with the nice level you can still opt to have it truncated, brew will just suggest the fully functional option.

Comment (dduugg) [reply Library/Homebrew/service.rb]: Could you elaborate on the issue in using `NICE_RANGE` in tests?

File: Library/Homebrew/service.rb [hunk 8]
```diff
@@ -355,6 +357,23 @@ def command
...
+      case T.unsafe(value)
+      when nil
+        @nice
+      when Integer
+        if NICE_RANGE.cover?(value)
+          @nice = value
+        else
+          raise TypeError,
+                "Service#nice value should be between #{NICE_RANGE}"
+        end
```
Comment (dduugg) [inline Library/Homebrew/service.rb:371]:
```suggestion
        raise TypeError, "Service#nice value should be in #{NICE_RANGE}" unless NICE_RANGE.cover?(value)

        @nice = value
```
As a slightly improved error message and slightly more idiomatic ruby

File: Library/Homebrew/test/service_spec.rb [hunk 9]
```diff
@@ -49,6 +49,34 @@ def stub_formula(&block)
     end
   end
 
+  describe "#nice" do
```
Comment (SMillerDev) [inline Library/Homebrew/test/service_spec.rb:52]: Can you add a `bool` check or something still?

File: Library/Homebrew/service.rb (hunk 6 above)
Comment (carlocab) [reply Library/Homebrew/service.rb]: Sounds like we should just have `nice < 0` imply `requires_root` then, regardless of whether the `service` block includes it. Not sure what the value of forcing its inclusion by throwing an error is.

Comment (SMillerDev) [reply Library/Homebrew/service.rb]: It would make it clear to developers and maintainers why it requires root. Instead of only telling users that they need to use root.

File: Library/Homebrew/test/service_spec.rb [hunk 10]
```diff
@@ -49,6 +49,34 @@ def stub_formula(&block)
...
+    it "throws for nice too low" do
+      f = stub_formula do
+        service do
+          run opt_bin/"beanstalkd"
+          nice(-21)
```
Comment (reitermarkus) [inline Library/Homebrew/test/service_spec.rb:57]:
```suggestion
          nice -21
```

Or does RuboCop complain about this?

File: Library/Homebrew/service.rb (hunk 6 above)
Comment (carlocab) [reply Library/Homebrew/service.rb]: I don't think an extra line that says `require_root true` is going to make the reason more transparent.

File: Library/Homebrew/test/service_spec.rb (hunk 10 above)
Comment (jorpilo) [reply Library/Homebrew/test/service_spec.rb:57]: yeah... syntax
